aspnetcore/test/dev-tunnels.md

```diff
@@ -0,0 +1,169 @@
+  * Select the URL bar and the QR code button appears.
+  * Select the QR code button to generate and view the QR code.
+    :::image type="content" source="~/test/dev-tunnels/_static/generate-qr-code.png" alt-text="QR code with button to create it highlighted.":::
+* Scan this QR code with a phone or tablet to navigate to the URL.
+
+## Dev Tunnels output window
+
+The **Dev Tunnels** output window shows the port and URL of a tunnel when a project starts using it. Select the **Dev Tunnels** output window by using the **Show output from:** dropdown in the **Output** window.
+
+:::image type="content" source="~/test/dev-tunnels/_static/dev-tunnels-output-window.png" alt-text="Dev Tunnels output window.":::
+
+This window is especially useful for projects that don't open a browser by default. For example, when working with an Azure Function, this may be the easiest way to discover the public URL that is being used by the dev tunnel.
+
+## Dev Tunnels tool window
+
+View and manage dev tunnels in the **Dev Tunnels** tool window:
+
+:::image type="content" source="~/test/dev-tunnels/_static/dev-tunnels-tool-window.png" alt-text="Dev Tunnels tool window.":::
+
+To open this window, Use the **Show Dev Tunnels Window** menu option in the debug dropdown. Or select **View** > **Other Windows** > **Dev Tunnels**.
+
+From this tool window, create a new tunnel by selecting the green `+` button. When you create a tunnel from the tool window, it isn't automatically set as the active tunnel like it is when a tunnel is created from the debug dropdown.
+
+Delete a tunnel by using the red `x` button on the right side.
+
+## Tunnel URL environment variable support
+
+The dev tunnels feature provides a way to get the tunnel URL of a project programmatically at run time. When you launch an app that uses a tunnel, Visual Studio creates an environment variable named `VS_TUNNEL_URL`. This is the URL for the tunnel that is used for the current project. It can be useful when you're integrating the application with an external service, and you need the URL to pass to the external service.
+
+If multiple ASP.NET Core projects are configured to start in Visual Studio, the app that is starting up will get an environment variable for any project that started before it. The pattern for this variable name is `VS_TUNNEL_URL_{ProjectName}`, where `{ProjectName}` is the name of the other project. For example, consider this example showing two projects set to start:
+
+:::image type="content" source="~/test/dev-tunnels/_static/multi-project-start.png" alt-text="Startup projects selection page showing MyWebApi and MyWebApp both starting, in that order.":::
+
+Since MyWebApi is above MyWebApp, it's started before the MyWebApp project. When the MyWebApi project is started, it will receive its tunnel URL in `VS_TUNNEL_URL`. When the MyWebApp project is started, it will receive its own tunnel URL in `VS_TUNNEL_URL` and the other project's tunnel URL in `VS_TUNNEL_URL_MyWebApi`. 
```

Comment:
```suggestion
Since MyWebApi is above MyWebApp, it's started before the MyWebApp project. When the MyWebApi project is started, it receives its tunnel URL in the `VS_TUNNEL_URL` environment variable. When the MyWebApp project is started, it receives its own tunnel URL in `VS_TUNNEL_URL` and the other project's tunnel URL is created in the `VS_TUNNEL_URL_MyWebApi` environment variable. 
```